src/kokkos_abstraction.hpp

```diff
@@ -214,476 +296,374 @@ inline void kokkos_dispatch(ParallelScanDispatch, Args &&...args) {
-  Kokkos::parallel_for(
-      name, team_policy(exec_space, NnNkNj, Kokkos::AUTO),
-      KOKKOS_LAMBDA(team_mbr_t team_member) {
-        int n = team_member.league_rank() / NkNj;
-        int k = (team_member.league_rank() - n * NkNj) / Nj;
-        int j = team_member.league_rank() - n * NkNj - k * Nj + jl;
-        n += nl;
-        k += kl;
-        Kokkos::parallel_for(Kokkos::TeamThreadRange<>(team_member, il, iu + 1),
-                             [&](const int i) { function(n, k, j, i); });
-      });
-}
+template <typename Pattern, typename... AllArgs>
+KOKKOS_FORCEINLINE_FUNCTION void par_disp_inner(Pattern, team_mbr_t team_member,
+                                                AllArgs &&...args) {
+  using dispatchsig = DispatchSignature<TypeList<AllArgs...>>;
+  constexpr std::size_t Rank = dispatchsig::Rank;
+  using Function = typename dispatchsig::Function;
+  using LoopBounds = typename dispatchsig::LoopBounds;
+  using Args = typename dispatchsig::Args;
+  using ExtraFuncArgs = typename function_signature<Rank, Function>::FArgs;
+  par_disp_inner_impl<Pattern, Function, LoopBounds, Args, ExtraFuncArgs>().execute(
+      team_member, std::forward<AllArgs>(args)...);
+}
+
+template <typename, typename, typename, typename, typename, typename>
+struct par_dispatch_impl {};
+
+template <typename Tag, typename Pattern, typename Function, typename... Bounds,
+          typename... Args, typename... ExtraFuncArgs>
+struct par_dispatch_impl<Tag, Pattern, Function, TypeList<Bounds...>, TypeList<Args...>,
+                         TypeList<ExtraFuncArgs...>> {
+  using LP = LoopPattern;
+  using dispatch_type = DispatchType<Tag, Pattern, Bounds...>;
+  using bound_translator = LoopBoundTranslator<Bounds...>;
+  static constexpr std::size_t Rank = bound_translator::Rank;
+
+  template <typename ExecSpace>
+  inline void dispatch(std::string name, ExecSpace exec_space, Bounds &&...bounds,
+                       Function function, Args &&...args, const int scratch_level = 0,
+                       const std::size_t scratch_size_in_bytes = 0) {
+    constexpr std::size_t Ninner =
+        dispatch_type::TeamPattern::Nvector + dispatch_type::TeamPattern::Nthread;
+    constexpr auto pattern_tag = LoopPatternTag<dispatch_type::GetPatternTag()>();
+    static_assert(
+        !std::is_same_v<decltype(pattern_tag), LoopPatternTag<LP::undef>> &&
+            !always_false<Tag, Pattern>,
+        "Loop pattern & tag combination not recognized in DispatchType::GetPatternTag");
+
+    constexpr bool isSimdFor =
+        std::is_same_v<LoopPatternSimdFor, base_type<decltype(pattern_tag)>>;
+    auto bound_arr = bound_translator().GetIndexRanges(std::forward<Bounds>(bounds)...);
+    if constexpr (isSimdFor) {
+      static_assert(!isSimdFor || (isSimdFor && std::is_same_v<ExecSpace, HostExecSpace>),
+                    "SimdFor pattern only supported in HostExecSpace");
+      SimdFor(std::make_index_sequence<Rank - 1>(), function, bound_arr);
```

Comment:
Is it possible to have a `dispatch_impl` overload for the simd for pattern instead of the `if` block here, just so that everything is uniform?

Reply:
I also did the same for the inner loops